### PR TITLE
env_aws: get ec2 cpu perf data from AWS API

### LINKED ISSUE #7830: Create a process to scrape EC2 CPU information out of AWS API
https://github.com/hashicorp/nomad/pull/7828 introduces a "best-effort" lookup table of CPU information per EC2 instance type.

This table is going to bitrot over time as Amazon introduces and upgrades instance types.

There is an EC2 [API](https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instance-types.html) available that looks to have the information we need
```
processor-info.sustained-clock-speed-in-ghz - The CPU clock speed, in GHz.
vcpu-info.default-cores - The default number of cores for the instance type.
vcpu-info.default-threads-per-core - The default number of threads per core for the instance type.
vcpu-info.default-vcpus - The default number of vCPUs for the instance type.
```

Using the API requires AWS secrets, which I think precludes the possibility of having Nomad Client lookup this data directly. 

Instead, we could establish some kind of process for generating a JSON version (or direct Go source file?) of the table using this API. E.g. part of the Nomad release process could be to push a button that triggers a pre-configured CI job that queries the API, generates the metadata table, and commits it to the nomad repo. Then `aws_env.go` would be redesigned to consume that generated content.

### PR DESCRIPTION
**Pull request by shoenig**
Previously, Nomad was using a hand-made lookup table for looking
up EC2 CPU performance characteristics (core count + speed = ticks).

This data was incomplete and incorrect depending on region. The AWS
API has the correct data but requires API keys to use (i.e. should not
be queried directly from Nomad).

This change introduces a lookup table generated by a docker image based
on the Amazon cli tool.

https://hub.docker.com/r/hashicorpnomad/ec2info

The docker image is created from https://github.com/hashicorp/nomad-ec2info.

Going forward, Nomad can incorporate regeneration of the lookup table
somewhere in the CI pipeline so that we remain up-to-date on the latest
offerings from EC2.

Fixes #7830

### CODE REVIEW

File: client/fingerprint/env_aws_cpuinfo.go [hunk 1]
```diff
@@ -0,0 +1,1620 @@
+// Code generated from hashicorpnomad/ec2info; DO NOT EDIT.
```
Comment (notnoop) [inline client/fingerprint/env_aws_cpuinfo.go:1]: Seeing that this is basically a shell script, I'd include it in the OSS repo with instructions.  One benefit might be getting contributions when the data is stale.

File: client/fingerprint/env_aws_cpuinfo.go [hunk 2]
```diff
@@ -0,0 +1,1620 @@
...
+func specs(ghz float64, cores int) ec2specs {
+	return ec2specs{
+		mhz:   ghz * 1000,
+		cores: cores,
+	}
+}
+
+var ec2ProcSpeedTable = map[string]map[string]ec2specs{
+
+	"ap-northeast-1": {
```
Comment (notnoop) [inline client/fingerprint/env_aws_cpuinfo.go:19]: Are we worried about specs for the same instance varying by region?  I'd find that very surprising if that's the case?

Looking at the current file, I don't see any divergence.  Deduplicating it reduces the elements from 1569 map entries to 212.

Also, I suspect some missing data: `c5a.16xlarge` and `c5.18xlarge` are available in multiple regions according to their press releases but they are only listed for eu-north-1 here: https://aws.amazon.com/blogs/aws/new-amazon-ec2-c5a-instances-powered-by-2nd-gen-amd-epyc-processors/, and https://aws.amazon.com/blogs/aws/now-available-compute-intensive-c5-instances-for-amazon-ec2/ -

**Review by notnoop — COMMENTED** [review]
Love the automation and code generation!  Few questions!

**Comment by shoenig** [comment]
@notnoop So yeah, I wrote a little tool to check if instance types ever vary by region, and (at least for now) they do not. 

I then cleaned up the tool so that it generates a minimal table and added it to Nomad's `tools` module under `nomad/tools/ec2info`. The tool uses the `aws-sdk-go` package and a template file to generate the output table, overwriting the previously generated output in `client/fingerprint`. 

There is no more Docker image or anything external, contributes now only need to set `AWS_*` and run
`go run .` in `nomad/tools/ec2info` to generate the table from fresh data.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ IMPROVEMENTS:
  * client: Added support for Azure fingerprinting. [[GH-8979](https://github.com/hashicorp/nomad/issues/8979)]
  * client: Added support for fingerprinting the client node's Consul segment. [[GH-7214](https://github.com/hashicorp/nomad/issues/7214)]
  * client: Updated consul-template to v0.25.0 - config function_blacklist deprecated and replaced with function_denylist [[GH-8988](https://github.com/hashicorp/nomad/pull/8988)]
+ * client: Use ec2 CPU perf data from AWS API [[GH-7830](https://github.com/hashicorp/nomad/issues/7830)]
```
Comment (tgross) [inline CHANGELOG.md:11]: changelog nitpick: this line should go above the longer lines above it.

**Review by tgross — APPROVED** [review]
LGTM 👍

**Pull request merged**